parallel.go

```diff
@@ -151,16 +178,28 @@ func (b *blockDesc) decompress() {
 	b.duration = time.Since(start)
 }
 
-func (dc *Decompressor) worker(ctx context.Context, in <-chan *blockDesc, out chan<- *blockDesc) {
+func (dc *Decompressor) worker(ctx context.Context, in <-chan *blockDesc, out chan<- *blockDesc, pool chan struct{}) {
 	for {
 		select {
 		case block := <-in:
 			if block == nil {
 				return
 			}
+			if pool != nil {
+				// Wait for a token from the pool.
+				select {
+				case <-pool:
+				case <-ctx.Done():
+					return
+				}
+			}
 			dc.trace("decompressing: %s", block)
 			block.decompress()
 			dc.trace("decompressed: %s, ch %v/%v", block, len(out), cap(out))
+			if pool != nil {
+				pool <- struct{}{}
```

Comment:
@cosnicolaou 

No. 1) This will never be blocking. 2) We must return the token, otherwise the system will get starved for tokens eventually. We pulled the token off, so there must be space for it.